### PR TITLE
Multithreading Issues and Memory Leaks

### PR DESCRIPTION
**Pull request by nuertey**
Changelog :
[1] Ensuring that resources (variables, threads, etc.) are created only at the point when needed.

[2] Ensuring that both threads cooperate better and in writing output to the Serial Terminal.

[3] Resolving memory leaks.

[4] Minor cleanups for code readability.

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -21,79 +21,98 @@
...
-    static int num = 0;
-    int count;
+    static size_t num = 0;
+    size_t count;
     
     memset(&stats[0], 0, sizeof(mbed_stats_socket_t) * MBED_CONF_NSAPI_SOCKET_STATS_MAX_COUNT);
     printf("%-15s%-15s%-15s%-15s%-15s%-15s%-15s\n", "Num", "ID", "State", "Proto", "Sent", "Recv", "Time");
-    while (1) {
-
+    while (COMPLETED_FLAG != threadFlag.get())
```
Comment (deepikabhavnani) [inline main.cpp]:
```suggestion
    while (COMPLETED_FLAG != threadFlag.get())  {
```
Please revert this formatting change (opening braces inside functions are on same line and new line only for functions), it is as per Mbed OS coding guidelines.

Comment (0xc0170) [reply main.cpp]: @nuertey use astyle and this Mbed OS AStyle configuration https://github.com/ARMmbed/mbed-os/blob/master/.astylerc

**Comment by deepikabhavnani** [comment]
@nuertey - This PR will need a rebase and minor formatting change.

**Comment by nuertey** [comment]
@deepikabhavnani , I reverted all those formatting changes as asked and pushed it to my fork. Let me know if I missed any or went through the wrong process to do so.

**Comment by nuertey** [comment]
Okay @deepikabhavnani, the formatting should be all taken care of now per the coding standard.

**Review by deepikabhavnani — APPROVED** [review]

**Comment by deepikabhavnani** [comment]
Some conflicts which should be resolved

**Comment by nuertey** [comment]
Essentially you want to remove the stdio_mutex then? Note that without the mutex, the output being issued by both threads becomes garbled on the serial terminal. That was one of the original fixes that called for the mutex. Do you want me to try removing that and then testing again to see if perhaps it is better now, what with all the other changes?

**Comment by deepikabhavnani** [comment]
> Essentially you want to remove the stdio_mutex then? 

No mutex is good. In order to resolve conflicts you will have to `rebase` your PR commits on top of upstream master to resolve conflicts. I will not be able to merge with conflicted code, in resolution you can select your code and pick master changes if missed by your commit
```

<<<<<<< master
        stdio_mutex.lock();
        printf("sent %d [%.*s]\n", result, strstr(sbuffer, "\r\n") - sbuffer, sbuffer);
        stdio_mutex.unlock();
=======
        printf("sent %d [%.*s]\n", result, strstr(sbuffer, "\r\n") - sbuffer, sbuffer);
>>>>>>> master

```
Will resolve into
```
        stdio_mutex.lock();
        printf("sent %d [%.*s]\n", result, strstr(sbuffer, "\r\n") - sbuffer, sbuffer);
        stdio_mutex.unlock();

```

**Comment by nuertey** [comment]
Thanks. I hope I got it now. I am fairly new to the github workflow so still learning some of the process as  I encounter them.

**Comment by deepikabhavnani** [comment]
> Thanks. I hope I got it now. I am fairly new to the github workflow so still learning some of the process as I encounter them.

Merge is another way of resolving conflicts. In mbed-os repo merges are not allowed and we suggest to do `git rebase` instead, you can explore that. But change is fine for this repo 👍

**Pull request merged**